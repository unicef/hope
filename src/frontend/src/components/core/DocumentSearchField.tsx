import { Grid, MenuItem } from '@mui/material';
import { SelectFilter } from '@core/SelectFilter';
import { SearchTextField } from '@core/SearchTextField';
import { ChoiceObject } from '@generated/graphql';
import { useTranslation } from 'react-i18next';

interface DocumentSearchFieldProps {
  onChange: (key: string, value: string) => void;
  type: string;
  number: string;
  choices: ChoiceObject[];
}

export const DocumentSearchField = ({
  onChange,
  type,
  number,
  choices = [],
}: DocumentSearchFieldProps) => {
  const { t } = useTranslation();

  return (
    <Grid container item xs={6} spacing={0}>
      <Grid item xs={4}>
        <SelectFilter
          onChange={(e) => onChange('documentType', e.target.value)}
          label={t('Document Type')}
          value={type}
          borderRadius="0px 4px 4px 0px"
          data-cy="filters-document-type"
          fullWidth
          disableClearable
        >
          {choices.map(({ name, value }) => (
            <MenuItem key={value} value={value}>
              {name}
            </MenuItem>
          ))}
        </SelectFilter>
      </Grid>
      <Grid item xs={8}>
        <SearchTextField
          value={number}
          label={t('Document Number')}
          placeholder="Document Number"
          onChange={(e) => onChange('documentNumber', e.target.value)}
          data-cy="filters-document-number"
          borderRadius="4px 0px 0px 4px"
        />
      </Grid>
    </Grid>
  );
};
