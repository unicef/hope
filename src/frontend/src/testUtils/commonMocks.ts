import { vi } from 'vitest';
import React from 'react';

/**
 * Common mock for @hooks/useBaseUrl hook
 */
export const mockUseBaseUrl = (
  businessArea = 'afghanistan',
  programId = 'test-program',
  baseUrl = 'afghanistan/test-program',
) => {
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({ businessArea, programId, baseUrl }),
  }));
};

/**
 * Common mock for useProgramContext hook
 */
export const mockUseProgramContext = (
  groupLabel = 'Household',
  groupLabelPlural = 'Households',
) => {
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel,
          groupLabelPlural,
        },
      },
    }),
  }));
};

/**
 * Common mock for react-router-dom
 */
export const mockReactRouterDom = () => {
  vi.mock('react-router-dom', () => ({
    useNavigate: () => vi.fn(),
    useLocation: () => ({
      pathname: '/test-path',
      search: '',
      hash: '',
      state: null,
      key: 'test-key',
    }),
    Link: ({ children, to, ...props }: any) =>
      React.createElement('a', { href: to, ...props }, children),
  }));
};
/**
 * Common mock for @utils/utils module
 */
export const utilsMock = {
  createApiParams: vi.fn((baseParams, queryParams, options) => ({
    ...baseParams,
    ...queryParams,
    ...options,
  })),
  adjustHeadCells: vi.fn((headCells) => headCells),
  formatCurrencyWithSymbol: vi.fn(
    (amount, currency) => `${amount} ${currency}`,
  ),
  householdStatusToColor: vi.fn(() => '#000000'),
  filterEmptyParams: vi.fn((params) => {
    if (!params) return {};
    return Object.fromEntries(
      Object.entries(params).filter(([, value]) => {
        return value !== undefined && value !== null && value !== '';
      }),
    );
  }),
  opacityToHex: vi.fn((opacity) =>
    Math.round(opacity * 255)
      .toString(16)
      .padStart(2, '0'),
  ),
  isPermissionDeniedError: vi.fn((error) => error?.response?.status === 403),
  dateToIsoString: vi.fn((date) => date?.toISOString?.() || date),
  choicesToDict: vi.fn((choices) => {
    if (!choices) return {};
    return choices.reduce((acc, choice) => {
      acc[choice.value] = choice.name;
      return acc;
    }, {});
  }),
  populationStatusToColor: vi.fn(() => 'primary'),
  sexToCapitalize: vi.fn(
    (sex) => sex?.charAt(0).toUpperCase() + sex?.slice(1).toLowerCase(),
  ),
  programStatusToColor: vi.fn(() => 'primary'),
  formatCurrency: vi.fn((amount) => `$${amount?.toLocaleString() || '0'}`),
  columnToOrderBy: vi.fn((column) => column || 'name'),
};

/**
 * Helper function to mock the @utils/utils module
 */
export const mockUtils = () => {
  vi.mock('@utils/utils', () => utilsMock);
};

/**
 * Common mock for RestService
 */
export const mockRestService = () => {
  vi.mock('@restgenerated/services/RestService', () => ({
    RestService: {
      restBusinessAreasProgramsHouseholdsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsCountRetrieve: vi.fn(),
      restBusinessAreasProgramsIndividualsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsMembersList: vi.fn(),
      restBusinessAreasProgramsList: vi.fn(),
      restBusinessAreasProgramsCountRetrieve: vi.fn(),
      restBusinessAreasUsersProfileRetrieve: vi.fn(),
      // Add other commonly used RestService methods as needed
    },
  }));
};

/**
 * Helper function to set up all common mocks at once
 * Usage in test files:
 *
 * import { setupCommonMocks } from 'src/testUtils/commonMocks';
 *
 * // At the top level of your test file
 * setupCommonMocks();
 */
export const setupCommonMocks = () => {
  // Mock useBaseUrl
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({
      businessArea: 'afghanistan',
      programId: 'test-program',
      baseUrl: 'afghanistan/test-program',
    }),
  }));

  // Mock useProgramContext
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel: 'Household',
          groupLabelPlural: 'Households',
        },
      },
    }),
  }));

  mockReactRouterDom();
  mockUtils();
  mockRestService();
};
