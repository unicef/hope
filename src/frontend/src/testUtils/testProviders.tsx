import { ThemeProvider } from '@mui/material';
import CssBaseline from '@mui/material/CssBaseline';
import * as React from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { theme } from '../theme';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';

interface TestProvidersProps {
  children: React.ReactNode;
}

export const TestProviders: React.FC<TestProvidersProps> = ({ children }) => (
  <ThemeProvider theme={theme}>
    <StyledThemeProvider theme={theme}>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <CssBaseline />
        {children}
      </LocalizationProvider>
    </StyledThemeProvider>
  </ThemeProvider>
);
