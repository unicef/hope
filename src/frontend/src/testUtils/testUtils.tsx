import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
