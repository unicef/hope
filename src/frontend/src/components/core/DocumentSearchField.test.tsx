import { render } from '../../testUtils/testUtils';
import { DocumentSearchField } from '@core/DocumentSearchField';

describe('components/core/DocumentSearchField', () => {
  it('should render', () => {
    const onChangeHandler = () => {};
    const choices = [
      { value: 'Document 1', name: 'Document 1' },
      { value: 'Document 2', name: 'Document 2' },
      { value: 'Document 3', name: 'Document 3' },
      { value: 'Document 4', name: 'Document 4' },
    ];

    const { container } = render(
      <DocumentSearchField
        onChange={onChangeHandler}
        type="Document 1"
        number="123123123"
        choices={choices}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
