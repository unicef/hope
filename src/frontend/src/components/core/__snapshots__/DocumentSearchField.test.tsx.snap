// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/core/DocumentSearchField should render 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6 css-137v0wv-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
    >
      <div
        class="sc-dntaoT dIKcKO"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth sc-blHHSb dhZMPX css-q8hpuo-MuiFormControl-root"
          data-cy="select-filter"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-c63xp4-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
            >
              Document Type
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall sc-ivxoEo eONSFc css-26qevy-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              data-cy="filters-document-type"
            >
              <div
                aria-controls=":r0:"
                aria-expanded="false"
                aria-haspopup="listbox"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                role="combobox"
                tabindex="0"
              >
                <div
                  class="MuiBox-root css-70qvj9"
                >
                  <div
                    class="sc-ghWlax crbcJj"
                  >
                    Document 1
                  </div>
                </div>
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="Document 1"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Document Type
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-vok43g-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root sc-kLhKbu LXzRU css-wb57ya-MuiFormControl-root-MuiTextField-root"
        data-cy="filters-document-number"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-c63xp4-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r1:"
          id=":r1:-label"
        >
          Document Number
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart css-11wfgpq-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-ittuaa-MuiInputAdornment-root"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
          </div>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart css-1ua80n0-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            maxlength="200"
            placeholder="Document Number"
            type="text"
            value="123123123"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Document Number
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
